kimchi/cairo/src/runner/bytecode.rs

```diff
@@ -0,0 +1,106 @@
+// TODO(querolita):
+// - be able to index memory with any type (u256 in particular) not only usize
+
+use crate::runner::word::CairoWord;
+
+/// This data structure stores the memory of the program
+pub struct CairoBytecode {
+    /// length of the public memory
+    publen: u64,
+    /// full memory vector, None if non initialized
+    pub vect: Vec<Option<CairoWord>>,
+}
+
+impl CairoBytecode {
+    /// Create a new memory structure from a vector of u64
+    pub fn new(input: Vec<i128>) -> CairoBytecode {
+        // Initialized with the public memory (compiled instructions only)
+        // starts intentionally with a zero word for ease of testing
+        let mut aux = vec![0];
+        aux.extend(input);
+        CairoBytecode {
+            publen: (aux.len() - 1) as u64,
+            vect: aux.into_iter().map(|i| Some(CairoWord::new(i))).collect(),
+        }
+    }
+
+    /// Get size of the public memory
+    pub fn public(&self) -> u64 {
+        self.publen
+    }
+
+    /// Get size of the full memory including dummy 0th entry
+    pub fn size(&self) -> u64 {
+        (self.vect.len()) as u64
+    }
+
+    /// Enlarges memory with enough additional None slots if necessary before writing or reading
+    fn enlarge(&mut self, index: u64) {
+        // if you want to access an index of the memory but its size is less or equal than this
+        if self.size() <= index {
+            // you will need to extend the vector with enough spaces (taking into account that
+            // vectors start by index 0 and size starts in 1)
+            let additional = index - self.size() + 1;
+            self.vect.reserve(additional.try_into().unwrap());
+            for _ in 0..additional {
+                // Consider CairoBytecode having Option<CairoWord> so one can have None here
+                self.vect.push(None);
+            }
+        }
+    }
+
+    /// Write u64 element in memory address
+    pub fn write(&mut self, index: u64, elem: i128) {
+        self.enlarge(index);
+        self.vect[index as usize] = Some(CairoWord::new(elem));
+    }
+
+    /// Read element in memory address
+    /// Because of how assignments work in Cairo (called assert-equal), they
+    /// behave in two ways: either check two variables are equal or assign
+    /// the value of one variable to the address of the other one so that
+    /// their content will be the same. This means you may first read to
+    /// addressses of memory that were still not instantiated, and you need
+    /// to enlarge the vector before reading (with None values).
+    pub fn read(&mut self, index: u64) -> Option<i128> {
+        self.enlarge(index);
+        if self.vect[index as usize].is_some() {
+            Some(self.vect[index as usize].unwrap().to_i128())
```

Comment:
always avoid `unwrap()`. Here write:

```rust
if let Some(x) = self.vect[index as usize] { ... }
```

or better:

```rust
self.vect[index as usize].map(|x| x.to_i128())
```

Reply:
Oh I see, so the `map` function would return an `Option<CairoWord>` meaning if the content of the address is a `None`, it will return a `None`... It looks way neater this way, it avoids using the conditional explicitly, thanks for the training! Nonetheless, what's the reason that `unwrap()` should be avoided? Perhaps the fact that it can panic if executed on a `None` element, and thus too verbose and non-elegant code?

Reply:
yeah! So in general you want to write code that can't panic, especially if it's library code. It's better to return an Error or an Option and let the caller decide what they want to do with that. If it doesn't make sense to return an Error/Option and let the caller decide, you can make the panic more explicit in several ways:

* use `expect()`, so that you can document why you think panicking is ok
* change the name of the function to reflect that. In OCaml functions that panic are appended with `_exn`, but I don't think there's a convention in Rust. Knowing that, I'm not sure if it makes sense to do that.
* document that the function can panic, this is what the rust stdlib does

Reply:
Thanks for the tips, I will be using them from now on 👍